### ISSUE TITLE
Mana creepers causing Memory leak

### ISSUE THREAD

**Issue by Ushupiuck** [Growlith1223/ArsMagica2#65]
As the tittle says. if I let a creeper explode on my face, my FPS will drop to 0, the "allocated memory" will get to 100%... nedless to mention, it will remain on 0-1 FPS... unless you open the options menu. (when you do, Fps get stable, but only while menu is open. as soon as you close the menu it drops back)
If this helps, let me mention that I'm using BetterFPS and Optifine.

Thanks for reading, and I hope this gets fixed... If needed, I can post a picture of my whole mod list.
PD.: I can't test without optifine; it becomes too laggy without it. (got 56 mods... including AM2)

**Comment by Zieg777** [comment]
Might be related to particles being weird lately.
A couple of my spells leave huge clouds of particles. Be good for a 3 second blindness I guess. Seems like it spawns too many particles, way too big too.

**Comment by Ushupiuck** [comment]
actually yes... that should be the issue; I had like a fog of particles at my feet when it exploded XD So many that my feet was covered entirely; it looked like RL fog. And, yeah, too big. Good luck with performance improvements!
Small posdata. You can always try latest forge.

**Comment by Zieg777** [comment]
https://gfycat.com/CanineResponsibleIrishredandwhitesetter

Behold a cloud maker!

**Comment by Ushupiuck** [comment]
Lol xD mine was similar Although condensed instead of dispersed... And only on the ground.
Off topic question, but do you plan some crossover items between this and blood magic?

**Comment by Growlith1223** [comment]
you never know, someone might want a fog machine!

real talk, i'll look into it, thanks! lol

**Comment by ProfHugo** [comment]
It looks like mana drain is generating this particle...

**Comment by Zieg777** [comment]
Makes sense. I get the same thing with my mana drain mana shield spell.

Related, Can we get a spell that does that cloud on purpose?

**Comment by Growlith1223** [comment]
you mean like, for aesthetic purposes or something?

**Comment by ProfHugo** [comment]
so.... mana clouds are a thing now?

**Comment by Growlith1223** [comment]
im not sure what's up with that memory leak, i can't reproduce it lol, i'll be posting a new download in a few so if you try that one and it persists, just lemme know and i'll check it out, would also help to know if you're using the mod alone or you have other mods as well

**Comment by Ushupiuck** [comment]
uhm... *points to first post* there you've got a mod list.... I listed the 2 I considered most important. if you need full mod list, however...
[this is it.](https://www.dropbox.com/s/cb6f1z8xl0kh1uv/mods.PNG?dl=0)
And I'll recall that I unfortunately can't test without optifine or betterFPS it becomes way too laggy if I do... (with the other mods installed and without optifine & betterFPS, I mean.)

**Comment by Ushupiuck** [comment]
@Growlith1223 No intend of bothering, but any advances so far? I mean this went silent, and if fixed, (I am unknown) never closed... just curious.

**Comment by ProfHugo** [comment]
tbh none of us were actually able to reproduce the memory leak by lighting up mana creepers. Just asking, but how much memory did you dedicated to java while running AM2? It's true that the "manafog" is very resource intensive, but it never caused an actual leak in my setup.

EDIT: The game starts to run out of memory after a few creepers got lit, but it recovered as the fog dissipates

**Comment by Ushupiuck** [comment]
Well... Actually I'm not sure if I used the correct term when saying "memory leak" but basically. yeah, the game ran out of memory for me, it got to 0 fps... I died instantly and got to spawn while the particles where still there. as for how much I got assigned..? My java arguments speak better than I.

-Xmx3100M -Xms256M -XX:+DisableExplicitGC -XX:+UseConcMarkSweepGC -XX:+UseParNewGC -XX:+UseNUMA -XX:+CMSParallelRemarkEnabled -XX:MaxTenuringThreshold=15 -XX:MaxGCPauseMillis=30 -XX:GCPauseIntervalMillis=150 -XX:+UseAdaptiveGCBoundary -XX:-UseGCOverheadLimit -XX:+UseBiasedLocking -XX:SurvivorRatio=8 -XX:TargetSurvivorRatio=90 -XX:MaxTenuringThreshold=15 -Dfml.ignorePatchDiscrepancies=true -Dfml.ignoreInvalidMinecraftCertificates=true -XX:+UseFastAccessorMethods -XX:+UseCompressedOops -XX:+OptimizeStringConcat -XX:+AggressiveOpts -XX:+UseCodeCacheFlushing -XX:SoftRefLRUPolicyMSPerMB=10000 -XX:ParallelGCThreads=10 -Xss2M

I would have said just ram, but I believe that chunk of argument I use is more useful... (I have 4GB of ram, and intel celeron dual core 1.58 Ghz each)

**Comment by ProfHugo** [comment]
I'm very sorry to say this, but...

By nature, your computer does lags a bit behind in performance, but with the proper settings it should be fine if nothing weird happens. You have a total of 4 gigs of available memory, but by dedicating 3.1 gigs to java, it left your system a total of 900 megs for everything else. My systems idles with 3 gigs (yea I have tons of stuff running in the background even right after bootup), and I guess it's safe to say that even if your laptop idles at 1-1.5 gigs like most computers do, your computer is on a spot where it doesn't have enough memory for anything else. That aside, due to how much memory the mana fog eats up, your computer was seriously starved on memory when the fog happened. 

When you go to debug mode, do u notice how the memory allocated would shifts from a small number all the way up to whatever ur XMX is? Well since your computer only has a grand total of 4 gigs, when it tries to chew more than it can bites on, that is 3.1 gigs, it can't use that much memory since with everything else running, the reserve is totally dry, causing a freeze (it can never use the memory you allocated to minecraft). For reference, I have 16 gigs of memory, a quarter of which is dedicated to MC, but even when the fog hits, since java can only use as much as I allowed it to with xmx, it wasn't much of a issue since there's still plenty for my computer to use. In short, you just ran out of memory.

A memory leak means that the memory that was used by a program doesn't freed up. For example, a program grabbed 1 gig of memory for it to use during its operation, Normally it should be able to return it to the pile so that it can be used again by other programs right? A memory leak happens when that program failed to return that chunk of memory it borrowed, making other programs unable to use it, thus resulting in the starvation of memory when the number racks up. 

tldr: Your computers can barely run minecraft, and this resource chewing fog just made it worse

**Comment by ProfHugo** [comment]
I guess it's better to just remove the fog from mana drain and life drain due to how resource intensive it is...

Well maybe with a bit of optimization a smoke screen component might be pretty fun

**Comment by Ushupiuck** [comment]
@ProfHugo Thanks for the advice! I'll see what I can I do on my side, after all, you're correct! Although, I need to mention, otherwise I might look a bit like an ignorant: Whenever I run MC, I always make sure that RAM is "clean"... (no other programs open, besies launcher some times, + a very lightweight & potent ram-cleaner program that is 1mb)

As for allocating the ram... I'll admit that I'm not even sure how much should be enough... vanilla can run normally with 256mb if I'm not mistaken (otherwise, maybe double? I don't recall if this has changed on latest version) but when modded? a total mistery, you have to find out by trial & error... To be honest, I used 3.1GB as max, as that fitted me perfectly on 1.7.10. more than enough... (on load, it can get to 95%, but after that? it "retreats" to 65-50%; which led me to believe it's fine)

**_PD.: Sorry for going off topic._**

**Comment by ProfHugo** [comment]
Normally I advice no more than 50% of your total ram. However, depending on the situation sometimes going toward the extreme side of the spectrum, well like 3.1/4 might not be such a bad call if you done your resource management well. Tbh I'm quite impress that you can get so many mods running with that thing...
